src/coreclr/src/vm/marshalnative.cpp

```diff
@@ -277,7 +277,7 @@ FCIMPL2(UINT32, MarshalNative::SizeOfClass, ReflectClassBaseObject* refClassUNSA
     // refClass is validated to be non-NULL RuntimeType by callers
     TypeHandle th = refClass->GetType();
 
-    if (throwIfNotMarshalable && !th.IsBlittable())
+    if (throwIfNotMarshalable && (!th.IsBlittable() || th.IsArray()))
```

Comment:
There is similar logic in `IsStructMarshalable`. Are these two related?

https://github.com/dotnet/runtime/blob/5a6c21cb6285a3c110f0048d9bf657042ea4ca10/src/coreclr/src/vm/fieldmarshaler.cpp#L311-L320

Reply:
Yes they are related. The `if` condition is basically a quick check for some possible failure cases so we can avoid doing a GC transition for blittable types.

Reply:
That should be noted here. Having these duplicate if checks is convoluted and breeds confusion about what is checking state and what isn't. I would like to see these two concept reconciled given the above has a nice comment but the one just introduced doesn't. This causes further confusion if one debugs through this a observes seemingly identical checks without a clear indicate as to why. Can the inner here be reused to fix this? If this is just an optimization to avoid the GC transition why was it still failing.